### PR TITLE
DBC-1400 - Merge diff of the latest fbs module.

### PR DESCRIPTION
**Pull request by NicolaeGrigoriu**

### CODE REVIEW

File: fbs.module [hunk 1]
```diff
@@ -213,15 +274,50 @@ function fbs_settings_form() {
     '#default_value' => variable_get('fbs_allow_profile_edit', FALSE),
   );
 
-  $form['fbs']['fbs_sort_departments'] = array(
```
Comment (guddo) [inline fbs.module]: @NicolaeGrigoriu Why is this removed?

File: fbs.module [hunk 2]
```diff
@@ -433,6 +552,19 @@ function fbs_interest_periods() {
   return $options;
 }
 
+/**
+ * Implements hook_form_FORM_ID_alter().
+ */
+function fbs_form_profile2_form_alter(&$form, &$form_state) {
+  $fee_sms = variable_get('ding_user_fee_sms', t('Notice that there is a fee for receiving a SMS'));
```
Comment (guddo) [inline fbs.module]: Does this `ding_user_fee_sms` exist?

File: fbs.module [hunk 3]
```diff
@@ -590,27 +749,44 @@ function _fbs_wayf_login($name, $value = NULL) {
 }
 
 /**
- * Implements hook_user_account_update().
```
Comment (guddo) [inline fbs.module]: `fbs_user_account_update()` not needed anymore?

File: includes/fbs.wayf.inc [hunk 4]
```diff
@@ -11,27 +11,17 @@
  * login.
  */
 function fbs_wayf_login_credentials($authentication_attributes) {
-  // Retrieve the CPR number from the UCFMD service.
```
Comment (guddo) [inline includes/fbs.wayf.inc]: This does nto seem right to remove..

File: src/ExternalCatalogApi.php [hunk 5]
```diff
@@ -12,78 +12,47 @@ class ExternalCatalogApi extends SwaggerApi
...
      * @param string $agencyid ISIL of the agency (e.g. DK-761500)
      * @param array $recordid list of record ids
-     * @return AvailabilityV3[]
+     * @param array $exclude Identifies the branchIds which are excluded from the result
+     * @return Availability[]
      */
-    public function getAvailability($agencyid, $recordid)
+    public function getAvailability($agencyid, $recordid, $exclude = null)
     {
-        $request = $this->newRequest("GET", "/external/{agencyid}/catalog/availability/v3");
```
Comment (guddo) [inline src/ExternalCatalogApi.php]: Why do we go from v3 to v2?

File: src/ExternalCatalogApi.php [hunk 6]
```diff
@@ -12,78 +12,47 @@ class ExternalCatalogApi extends SwaggerApi
...
+        $request = $this->newRequest("GET", "/external/v2/{agencyid}/catalog/availability");
         $request->addParameter("path", "agencyid", $agencyid);
         $request->addParameter("query", "recordid", $recordid);
+        $request->addParameter("query", "exclude", $exclude);
 
-        $request->defineResponse(200, "", array('\\FBS\\Model\\AvailabilityV3'));
```
Comment (guddo) [inline src/ExternalCatalogApi.php]: From v3 to v1?

File: fbs.features.field_instance.inc [hunk 7]
```diff
@@ -246,10 +241,9 @@ function fbs_field_default_field_instances() {
...
         'label' => 'above',
-        'module' => 'fbs',
         'settings' => array(),
-        'type' => 'fbs_preferred_branch',
-        'weight' => 7,
+        'type' => 'hidden',
+        'weight' => 2,
       ),
     ),
     'entity_type' => 'profile2',
```
Comment (AnatolyMuntean) [inline fbs.features.field_instance.inc]: There are a number of fields that are related to he user profile that got "hidden", like prefferred branch, interest period. etc. I assume these fields won't be displayed for the end user that tries to edit his profile. This might not be what is expected and must be checked with a test user.

File: fbs.field.inc [hunk 8]
```diff
@@ -260,6 +276,16 @@ function fbs_field_metadata_get($entity, array $options, $name, $entity_type, &$
   return NULL;
 }
 
+/**
+ * Callback to set the value of a field.
+ */
+function fbs_field_metadata_set($entity, $name, $items, $langcode, $entity_type) {
+  if (!isset($items['value'])) {
+    $items = array('value' => $items);
```
Comment (AnatolyMuntean) [inline fbs.field.inc]: Existence of a certain key value is performed, however whole `$items` array is overriden.
Would suggest to assign default value to the actual key that is checked.

File: includes/fbs.availability.inc [hunk 9]
```diff
@@ -12,29 +11,21 @@
...
-        $service->agencyId,
-        $provider_ids
-      );
-    }
-    catch (Exception $e) {
-      watchdog_exception('fbs', $e);
-    }
+
+  try {
+    $avails = fbs_service()->Catalog->getAvailability(fbs_service()->agencyId, $provider_ids, variable_get('fbs_holdings_branches_blacklist', array()));
```
Comment (AnatolyMuntean) [inline includes/fbs.availability.inc]: Hungarian notation should be avoided.

File: includes/fbs.availability.inc [hunk 10]
```diff
@@ -12,29 +11,21 @@
...
+    watchdog_exception('fbs', $e);
   }
 
   $result = array();
   foreach ($avails as $item) {
-    $result[$item->recordId] = array(
-      'available' => $item->available,
-      'reservable' => $item->reservable,
-    );
+    $result[$item->recordId] = array('available' => $item->available, 'reservable' => $item->reservable);
```
Comment (AnatolyMuntean) [inline includes/fbs.availability.inc]: Certain places have each array item on it's own line, others not.
Having multiple array values on each line helps readability.

File: includes/fbs.availability.inc [hunk 11]
```diff
@@ -51,37 +42,36 @@ function fbs_availability_items($provider_ids) {
...
-        $service->agencyId,
-        $provider_ids
-      );
-    }
-    catch (Exception $e) {
-      watchdog_exception($e);
-    }
+
+  try {
+    $holdings = fbs_service()->Catalog->getHoldings(fbs_service()->agencyId, $provider_ids, variable_get('fbs_holdings_branches_blacklist', array()));
```
Comment (AnatolyMuntean) [inline includes/fbs.availability.inc]: Looks messy at first. Improving readability is adviced:
```
fbs_service()->Catalog->getHoldings($agency_id, $provider_ids, $blacklist);
```

File: includes/fbs.debt.inc [hunk 12]
```diff
@@ -5,31 +5,6 @@
...
-/**
- * Fetches user debts.
- *
- * @param $account
- *   User object whose debts are fetched.
- *
- * @return array
- *   Set of user debts.
- */
-function fbs_debt_raw_list(stdClass $account) {
```
Comment (AnatolyMuntean) [inline includes/fbs.debt.inc]: This was specially added to be able to get debts list without any processing. Since invoking `fbs_debts_list()` also preloads entities which in case of use pages dramatically increases page load.

File: includes/fbs.reservation.inc [hunk 13]
```diff
@@ -345,31 +369,3 @@ function fbs_reservation_pickup_branches() {
 
   return $branches;
 }
-
```
Comment (guddo) [inline includes/fbs.reservation.inc]: Is this handled elsewhere?

File: src/ExternalAuthenticationApi.php [hunk 14]
```diff
@@ -30,7 +30,7 @@ public function login($agencyid, Model\Login $login)
         $request->addParameter("body", "login", $login);
 
         $request->defineResponse(200, "", '\\FBS\\Model\\ExternalAPIUserInfo');
-        $request->defineResponse("400", 'bad request', '\\FBS\\Model\\RestException');
+      $request->defineResponse("400", 'bad request', '\\FBS\\Model\\RestException');
```
Comment (guddo) [inline src/ExternalAuthenticationApi.php]: ?

File: src/Model/PatronV3.php [hunk 15]
```diff
@@ -1,102 +1,102 @@
...
-  public $defaultInterestPeriod = null;
-
-  /**
-   * @var boolean True if the user is resident in the same municipality as the
-   * library
-   * @required
-   */
-  public $resident = null;
-}
+<?php
```
Comment (guddo) [inline src/Model/PatronV3.php]: Why is there a diff here?

File: src/Model/UpdatePatronRequestV3.php [hunk 16]
```diff
@@ -1,16 +1,12 @@
...
-   * @var PatronSettingsV3 Set this if patron details are to be changed
-   */
-  public $patron = null;
-
-  /**
-   * @var PincodeChange Set this if pincode is to be changed
-   */
-  public $pincodeChange = null;
-}
+<?php
```
Comment (guddo) [inline src/Model/UpdatePatronRequestV3.php]: Why is there a diff here?

File: src/Model/PatronV3.php (hunk 15 above)
Comment (NicolaeGrigoriu) [reply src/Model/PatronV3.php]: This is code from the easySuite/ding2/fbs.

Comment (guddo) [reply src/Model/PatronV3.php]: But there is no real difference in code, it seems like your editor did something..
Btw. why easySuite/ding2 and not ding2/ding2?
